File: mobile/lib/utils/resolve_profile_media_url.dart
import 'dart:io';

import '../config/api_config.dart';

/// Resolve profile/cover media paths for display — mirrors webapp `resolveProfileMediaUrl`.
String resolveProfileMediaUrl(String? raw) {
  final trimmed = raw?.trim();
  if (trimmed == null || trimmed.isEmpty) return '';

  if (trimmed.startsWith('http://') ||
      trimmed.startsWith('https://') ||
      trimmed.startsWith('data:')) {
    return _normalizeHostForPlatform(trimmed);
  }

  final base = resolveApiBaseUrl();
  final path = trimmed.startsWith('/') ? trimmed : '/$trimmed';
  return _normalizeHostForPlatform('$base$path');
}

/// Android emulator cannot reach host `localhost` — use `10.0.2.2`.
String _normalizeHostForPlatform(String url) {
  if (!Platform.isAndroid) return url;
  final uri = Uri.tryParse(url);
  if (uri == null) return url;
  final host = uri.host.toLowerCase();
  if (host != 'localhost' && host != '127.0.0.1') return url;
  return uri.replace(host: '10.0.2.2').toString();
}

File: admin/src/lib/users/resolveProfileMediaUrl.ts
import { getApiOrigin } from '@/lib/api';
import { diceBearAvatarUrl, isDiceBearStoredProfile } from '@/lib/users/diceBearAvatarUrl';

/** Resolve stored profile image paths for admin display (matches webapp behavior). */
export function resolveProfileMediaUrl(
  raw: string | null | undefined,
  username: string
): string {
  const trimmed = raw?.trim();
  const handle = username.trim() || 'user';

  if (!trimmed || isDiceBearStoredProfile(trimmed)) {
    return diceBearAvatarUrl(handle);
  }

  if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) {
    if (isDiceBearStoredProfile(trimmed)) {
      return diceBearAvatarUrl(handle);
    }
    return trimmed;
  }

  if (trimmed.startsWith('data:')) {
    return trimmed;
  }

  const base = getApiOrigin().replace(/\/$/, '');
  const path = trimmed.startsWith('/') ? trimmed : `/${trimmed}`;
  return base ? `${base}${path}` : path;
}
